src/protocol/ipa_prf/boolean_ops/share_conversion_aby.rs

```diff
@@ -0,0 +1,361 @@
+        Role::H3 => {
+            sh_r.1 = <B as WeakSharedValue>::ZERO;
+            sh_s.0 = <B as WeakSharedValue>::ZERO;
+        }
+    }
+
+    let sh_rs = integer_add(
+        ctx.narrow(&Step::IntegerAddBetweenMasks),
+        record_id,
+        &sh_r,
+        &sh_s,
+    )
+    .await?;
+
+    let sh_y = integer_add(ctx.narrow(&Step::IntegerAddMaskToX), record_id, &sh_rs, x).await?;
+
+    let y = sh_y
+        .partial_reveal(ctx.narrow(&Step::RevealY), record_id, Role::H3)
+        .await?;
+
+    match ctx.role() {
+        Role::H1 => Ok(AdditiveShare::<A>(
+            A::from(sh_s.0).neg(),
+            A::from(y.unwrap()),
+        )),
+        Role::H2 => Ok(AdditiveShare::<A>(
+            A::from(y.unwrap()),
+            A::from(sh_r.1).neg(),
+        )),
+        Role::H3 => Ok(AdditiveShare::<A>(
+            A::from(sh_r.0).neg(),
+            A::from(sh_s.1).neg(),
+        )),
+    }
+}
+
+/// similar to convert however converting Boolean arrays to `Fp25519` takes special treatment
+/// requires `BA256` (to handle carries),
+/// can only be used securely to convert `BAt` to `Fp25519`,
+/// where t < 256 - statistical security parameter due to leakage
+///
+/// leakage free alternative needs secure mod p operation after each addition
+/// (which can be performed using secure subtraction)
+///
+/// we use a `BA256` for masks r, s and set the 2 most significant bits to 0.
+/// this allows us to compute Boolean shares of r + s without reducing it mod 256
+/// further it allows us to compute x + r + s without reducing it mod 256
+///
+/// We can then reveal x+r+s via `partial_reveal` and reduce it with mod p, where p is the prime of `Fp25519`
+///
+/// in the process, we need to make sure that highest order `PRSS` masks added by `multiply`
+/// are set to zero since they would also cause carries in the integer addition
+///
+/// Setting the most significant bits to 0 leaks information about x:
+/// assuming x has m bits, revealing y = x + rs (where rs = r + s) leaks the following information
+/// (see `bit_adder` in `protocol::ipa_prf::boolean_ops::addition_low_com::integer_add`):
+/// y{m} := rs{m} xor x{m} xor `carry_x`
+/// where x{m} is 0 and `carry_x` is carry{m-1} which contains information about x
+/// carry is defined as carry{i} = carry{i-1} xor (x{i-1} xor carry{i-1})(y{i-1} xor carry{i-1})
+/// for all i where x{i-1}=0: carry{i} = carry{i-1}y{i-1}
+/// therefore for j>=0,
+/// y{m+j} := rs{m+j} xor (`carry_x` * `product_(k=m)^(m+j-1)rs{k})`
+/// this will result in leakage (rs{255}=0, rs{254}=0)
+/// y{255} := `carry_x` * `product_(k=m)^(255)rs{k})`
+/// y{254} := `carry_x` * `product_(k=m)^(254)rs{k})`
+/// however, these terms are only non-zero when all rs{k} terms are non-zero
+/// this happens with probability 1/(2^(256-m)) which is negligible for small m
+/// # Errors
+/// Propagates Errors from Interger Subtraction and Partial Reveal
+#[cfg(all(test, unit_test))]
+async fn convert_to_fp25519<C, B>(
```

Comment:
This is a lot of duplicate code between this and `async fn convert<C, B, A>`. I wonder if there's a better way here to support both without so much duplication.